### ISSUE TITLE
globbing in FTP.ftp_NLST may be incomplete

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4181]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)| @jesstess reported|
|-|-|
|Trac ID|trac#4181|
|Type|defect|
|Created|2009-12-21 18:32:07Z|
|Branch|https://github.com/twisted/twisted/tree/ftp-nlst-globs-4181|

Attachments:

* [4181-nlst-globbing-1.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/5cb/5cb8fd377a6e0d2776fa766632f6a31060937156/c687d06a9fb445b5198ee0e0ddc60c2c90d45014.diff) (5643 bytes) - added by adiroiban on 2012-07-22 14:19:03Z - 
* [4181-nlst-globbing-2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/5cb/5cb8fd377a6e0d2776fa766632f6a31060937156/5c45c2d1d13d775a805c3bffb987cfe44a210b46.diff) (6412 bytes) - added by adiroiban on 2012-11-18 17:05:04Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4181 4181
type__defect defect
reporter__jesstess jesstess
priority__normal normal
milestone__ 
branch__branches_ftp_nlst_globs_4181 branches/ftp-nlst-globs-4181
branch_author__exarkun__tenth exarkun, tenth
status__closed closed
resolution__fixed fixed
component__ftp ftp
keywords__ 
time__1261420327000000 1261420327000000
changetime__1358899158000000 1358899158000000
version__None None
owner__exarkun exarkun
cc__jesstess cc__itamar cc__adiroiban
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst commented|
|-|-|

Unfortunately things like wildcards and "-a" are not part of FTP standard, but artifact of early FTP servers using /bin/ls to list files. Did you have a specific limitation in mind?

http://cr.yp.to/ftp/list.html is a handy reference to what is supported in practice (and of real world FTP in general). Looks like adding EPLF (a decent file listing command) may be worthy of another ticket.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to spiv|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban commented|
|-|-|

Hi,

I have created an 

The diff can be viewed here:
https://github.com/adiroiban/twisted/compare/4181-nlst-globbing

Please let me know if you want me to attach a svn diff.

----

The changes include an utility function for checking if a segment contains a globbing expression.

----

I don't know if globbing is used in NLST so maybe you can chose to fix this ticket by removing globbing functionality from NLST command.

----

Please let me know what is the next step for having this ticket solved.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @adiroiban|
|-|-|

[ticket:4180#comment:5 Same comment]

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_8"></a>|@adiroiban removed owner|
|-|-|

I have attached the diff file and I am not waiting for the review.

Thanks,
Adi

**Comment by twisted-trac** [comment]
|[<img alt="tenth10th's avatar" src="https://avatars.githubusercontent.com/u/854235?s=50" width="50" height="50">](https://github.com/tenth10th)<a name="note_9"></a>|@tenth10th commented|
|-|-|

(In [36381]) Branching to 'nlst-globbing-4181'

**Comment by twisted-trac** [comment]
|[<img alt="tenth10th's avatar" src="https://avatars.githubusercontent.com/u/854235?s=50" width="50" height="50">](https://github.com/tenth10th)<a name="note_10"></a>|@tenth10th commented|
|-|-|

Thanks for the patch!

A few concerns:

 1. As per Twisted's [Compatibility Policy](https://twistedmatrix.com/trac/wiki/CompatibilityPolicy), it's best to avoid public API additions unless they're necessary; isGlobbingExpression should probably be private.

 2. There seems to be some confusion between the code and the docstring: the argument to isGlobbingExpression is called "segments", but is referred to as "expression" in the docs. Ideally, the docstring should clearly state the type of the argument.

 3. It's not really clear what "globbing" is from the isGlobbingExpression docstring; even just a short explanation or link to additional docs would be helpful to future readers of the code. (I didn't really follow it until I read the test docstrings as well.)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph set owner to @adiroiban|
|-|-|

Re-assigning so Adi can respond; thanks for all this work on FTP, we really appreciate your dedication :).

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_12"></a>|@adiroiban set owner to @tenth10th|
|-|-|

Hi,

Many thanks for the review.

I have attached a new diff.

It renames isGlobbingExpression to _isGlobbingExpression

I hope that the new docstrings are helpful.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

(In [36884]) Branching to 'ftp-nlst-globs-4181'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

Thanks.  This looks pretty good.  It needs a news file and a few doc adjustments to follow our doc conventions.  I can make these simple changes and apply though.  Thanks again.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun set status to `closed`|
|-|-|

(In [36888]) Merge ftp-nlst-globs-4181

Author: adiroiban
Reviewer: tenth, exarkun
Fixes: [#4181](https://github.com/twisted/twisted/issues/4181)

More completely support globbing expressions in the FTP server's handling of the NLST command
by using `fnmatch` to detect globs, rather than having a custom, incomplete check for some subset
of the globbing syntax.
